File: apps/android/app/src/main/java/com/universalsim/extender/Account.kt
package com.universalsim.extender

import android.content.Context
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder

/**
 * A Universal ID on the phone — sign in, stay signed in, and let the saved
 * machines follow the account (James, 2026-09-17: "go ahead with screens
 * account", choosing "Identity + saved hosts follow you").
 *
 * The browser client's `apps/web/src/account.js` is the same thing in the same
 * order; keep the two in step. GoTrue's REST API written out by hand, because
 * Screens has no `@unisim/sdk` on any client:
 *
 *   POST /auth/v1/otp     { email, create_user: false }   → code emailed
 *   POST /auth/v1/verify  { email, token, type: email }   → session
 *   POST /auth/v1/token?grant_type=refresh_token          → fresh session
 *
 * **LOGIN ONLY.** A Universal ID is created on the hub (app.unisim.co.uk),
 * never here, so a typo cannot mint a stray account.
 *
 * ⚠️ **The PIN is never uploaded.** `SavedConnection` on this client DOES hold
 * a pin — unlike the browser's — and `screens_saved_hosts` (migration 0178) has
 * no column for it, deliberately: it is the Noise pre-shared key. Every mapping
 * below names its fields one by one for exactly that reason. Do not replace one
 * with a whole-object serialiser. `mode` and `hidden` are this client's own and
 * stay local too.
 *
 * ⚠️ Signing in is optional and changes nothing about connecting. Screens is
 * used on networks with no route out as a matter of course, so every call here
 * is best-effort and silent.
 */
object Account {
    private const val SUPABASE_URL = "https://rygfxgalojojppxmhddo.supabase.co"
    private const val SUPABASE_ANON =
        "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6InJ5Z2Z4Z2Fsb2pvanBweG1oZGRvIiwicm9sZSI6ImFub24iLCJpYXQiOjE3Nzg3NTY4MjUsImV4cCI6MjA5NDMzMjgyNX0.hLy_vt9vY_rdPKF3nL32yAuMCD604E3CH5VM7D7CaNE"

    private const val PREFS = "unisim_suite"
    private const val KEY_ACCESS = "auth:access"
    private const val KEY_REFRESH = "auth:refresh"
    private const val KEY_EXPIRES = "auth:expires"
    private const val KEY_EMAIL = "auth:email"
    private const val KEY_USER = "auth:user"

    private fun prefs(context: Context) = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

    /** The signed-in address, or null. Reads storage only. */
    fun email(context: Context): String? =
        prefs(context).getString(KEY_EMAIL, null)?.takeIf { it.isNotBlank() }

    fun userId(context: Context): String? =
        prefs(context).getString(KEY_USER, null)?.takeIf { it.isNotBlank() }

    private fun store(context: Context, session: JSONObject?) {
        val e = prefs(context).edit()
        if (session == null) {
            e.remove(KEY_ACCESS).remove(KEY_REFRESH).remove(KEY_EXPIRES).remove(KEY_EMAIL).remove(KEY_USER)
        } else {
            val user = session.optJSONObject("user")
            e.putString(KEY_ACCESS, session.optString("access_token"))
                .putString(KEY_REFRESH, session.optString("refresh_token"))
                // A minute of slack, so nothing is sent with a token about to expire.
                .putLong(KEY_EXPIRES, System.currentTimeMillis() + (session.optLong("expires_in", 3600) - 60) * 1000)
            if (user != null) {
                e.putString(KEY_EMAIL, user.optString("email")).putString(KEY_USER, user.optString("id"))
            }
        }
        e.apply()
    }

    /** POST to GoTrue. Returns the body, or throws with GoTrue's own message. */
    private fun auth(path: String, body: String): JSONObject {
        val conn = (URL("$SUPABASE_URL/auth/v1/$path").openConnection() as HttpURLConnection).apply {
            requestMethod = "POST"
            connectTimeout = 15_000
            readTimeout = 15_000
            doOutput = true
            setRequestProperty("apikey", SUPABASE_ANON)
            setRequestProperty("Content-Type", "application/json")
        }
        conn.outputStream.use { it.write(body.toByteArray()) }
        val ok = conn.responseCode in 200..299
        val text = (if (ok) conn.inputStream else conn.errorStream)?.bufferedReader()?.use { it.readText() } ?: ""
        conn.disconnect()
        val json = runCatching { JSONObject(text) }.getOrDefault(JSONObject())
        if (!ok) {
            val msg = json.optString("error_description").ifBlank { json.optString("msg") }
                .ifBlank { json.optString("message") }.ifBlank { "Sign-in failed (${conn.responseCode})" }
            throw IllegalStateException(msg)
        }
        return json
    }

    /** Email a 6-digit code to an existing Universal ID. */
    suspend fun sendCode(email: String) = withContext(Dispatchers.IO) {
        try {
            auth("otp", JSONObject().put("email", email.trim()).put("create_user", false).toString())
            Unit
        } catch (e: IllegalStateException) {
            // With create_user false, GoTrue answers an unknown address with
            // "Signups not allowed for otp". Say what that means here.
            if (e.message?.contains("not allowed", ignoreCase = true) == true) {
                throw IllegalStateException(
                    "No Universal ID for that email. Create one free at app.unisim.co.uk, then sign in here.",
                )
            }
            throw e
        }
    }

    /** Exchange the emailed code for a session. */
    suspend fun verifyCode(context: Context, email: String, code: String) = withContext(Dispatchers.IO) {
        val session = auth(
            "verify",
            JSONObject().put("email", email.trim()).put("token", code.trim()).put("type", "email").toString(),
        )
        if (session.optString("access_token").isBlank()) throw IllegalStateException("That code was not accepted.")
        store(context, session)
    }

    /** A usable access token, refreshing when due. Null when signed out. */
    suspend fun accessToken(context: Context): String? = withContext(Dispatchers.IO) {
        val p = prefs(context)
        val refresh = p.getString(KEY_REFRESH, null) ?: return@withContext null
        val access = p.getString(KEY_ACCESS, null)
        if (access != null && p.getLong(KEY_EXPIRES, 0) > System.currentTimeMillis()) return@withContext access
        try {
            val next = auth("token?grant_type=refresh_token", JSONObject().put("refresh_token", refresh).toString())
            store(context, next)
            next.optString("access_token").takeIf { it.isNotBlank() }
        } catch (_: Exception) {
            // A refused refresh means the session is gone for good (signed out
            // elsewhere, account deleted) — don't keep a dead one.
            store(context, null)
            null
        }
    }

    suspend fun signOut(context: Context) = withContext(Dispatchers.IO) {
        val token = prefs(context).getString(KEY_ACCESS, null)
        store(context, null)
        if (token != null) {
            runCatching {
                val conn = (URL("$SUPABASE_URL/auth/v1/logout").openConnection() as HttpURLConnection).apply {
                    requestMethod = "POST"
                    connectTimeout = 10_000
                    setRequestProperty("apikey", SUPABASE_ANON)
                    setRequestProperty("Authorization", "Bearer $token")
                }
                conn.responseCode
                conn.disconnect()
            }
        }
        Unit
    }

    /** An authorised REST call. Null when signed out or the call fails. */
    private fun rest(token: String, path: String, method: String = "GET", body: String? = null, prefer: String? = null): String? =
        runCatching {
            val conn = (URL("$SUPABASE_URL/rest/v1/$path").openConnection() as HttpURLConnection).apply {
                requestMethod = method
                connectTimeout = 15_000
                readTimeout = 15_000
                setRequestProperty("apikey", SUPABASE_ANON)
                setRequestProperty("Authorization", "Bearer $token")
                setRequestProperty("Content-Type", "application/json")
                prefer?.let { setRequestProperty("Prefer", it) }
                if (body != null) doOutput = true
            }
            if (body != null) conn.outputStream.use { it.write(body.toByteArray()) }
            val ok = conn.responseCode in 200..299
            val text = if (ok) conn.inputStream.bufferedReader().use { it.readText() } else null
            conn.disconnect()
            text
        }.getOrNull()

    /** The five fields that may leave the device. NOT pin, mode or hidden. */
    private fun row(userId: String, c: SavedConnection): JSONObject = JSONObject()
        .put("user_id", userId)
        .put("addr", c.addr)
        .put("hostname", c.hostname)
        .put("os", c.os)
        .put("custom_name", c.customName)
        .put("last_connected", iso(c.lastConnected))

    private fun iso(ms: Long): String {
        val millis = if (ms > 0) ms else System.currentTimeMillis()
        return java.time.Instant.ofEpochMilli(millis).toString()
    }

    /**
     * Merge this phone's saved machines with the account's — most recently
     * connected wins per address — write back whatever the account was missing,
     * and save the result. Returns true when the list changed.
     *
     * ⚠️ A remote row never overwrites this phone's `pin`, `mode` or `hidden`:
     * they are not synced, and losing the pin would mean retyping it to
     * reconnect to a machine the phone already knew.
     */
    suspend fun syncSaved(context: Context): Boolean = withContext(Dispatchers.IO) {
        val token = accessToken(context) ?: return@withContext false
        val userId = userId(context) ?: return@withContext false
        val body = rest(token, "screens_saved_hosts?select=addr,hostname,os,custom_name,last_connected")
            ?: return@withContext false

        val local = ConnectionStore.load(context)
        val merged = LinkedHashMap<String, SavedConnection>()
        local.forEach { merged[it.addr] = it }

        runCatching {
            val arr = JSONArray(body)
            for (i in 0 until arr.length()) {
                val o = arr.getJSONObject(i)
                val addr = o.getString("addr")
                val remoteAt = runCatching { java.time.Instant.parse(o.getString("last_connected")).toEpochMilli() }
                    .getOrDefault(0L)
                val mine = merged[addr]
                if (mine == null) {
                    merged[addr] = SavedConnection(
                        addr = addr,
                        hostname = o.optString("hostname"),
                        os = o.optString("os"),
                        customName = o.optString("custom_name"),
                        lastConnected = remoteAt,
                    )
                } else if (remoteAt > mine.lastConnected) {
                    // Labels from the account, pin/mode/hidden kept from here.
                    merged[addr] = mine.copy(
                        hostname = o.optString("hostname").ifBlank { mine.hostname },
                        os = o.optString("os").ifBlank { mine.os },
                        customName = o.optString("custom_name"),
                        lastConnected = remoteAt,
                    )
                }
            }
        }

        // Whatever the account did not have, or had staler than this phone.
        val remoteAddrs = runCatching {
            val arr = JSONArray(body)
            (0 until arr.length()).associate {
                val o = arr.getJSONObject(it)
                o.getString("addr") to (runCatching { java.time.Instant.parse(o.getString("last_connected")).toEpochMilli() }.getOrDefault(0L))
            }
        }.getOrDefault(emptyMap())
        val push = local.filter { (remoteAddrs[it.addr] ?: -1) < it.lastConnected }
        if (push.isNotEmpty()) {
            val arr = JSONArray()
            push.forEach { arr.put(row(userId, it)) }
            rest(
                token, "screens_saved_hosts?on_conflict=user_id,addr",
                method = "POST", body = arr.toString(), prefer = "resolution=merge-duplicates",
            )
        }

        val result = merged.values.sortedByDescending { it.lastConnected }
        ConnectionStore.replaceAll(context, result)
        result != local
    }

    /** Save (or refresh) one machine against the account. No-op signed out. */
    suspend fun remember(context: Context, addr: String) = withContext(Dispatchers.IO) {
        val token = accessToken(context) ?: return@withContext
        val userId = userId(context) ?: return@withContext
        val c = ConnectionStore.load(context).find { it.addr == addr } ?: return@withContext
        rest(
            token, "screens_saved_hosts?on_conflict=user_id,addr",
            method = "POST", body = JSONArray().put(row(userId, c)).toString(),
            prefer = "resolution=merge-duplicates",
        )
        Unit
    }

    /** Drop one machine from the account. No-op signed out. */
    suspend fun forget(context: Context, addr: String) = withContext(Dispatchers.IO) {
        val token = accessToken(context) ?: return@withContext
        rest(token, "screens_saved_hosts?addr=eq.${URLEncoder.encode(addr, "UTF-8")}", method = "DELETE")
        Unit
    }
}

/**
 * For the fire-and-forget mirrors — remembering or forgetting a machine against
 * the account as the user does it here. Deliberately NOT a composable's scope:
 * the write must not be cancelled because the screen it started from went away,
 * and `SupervisorJob` keeps one failed call from taking the next one with it.
 */
val accountScope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

File: apps/web/src/account.js
// A Universal ID in the browser client — sign in, stay signed in, and let the
// saved-machine list follow the account (James, 2026-09-17: "go ahead with
// screens account", choosing "Identity + saved hosts follow you").
//
// Screens has no `@unisim/sdk` on any client and no bundler on this one, so
// this is GoTrue's REST API written out directly. The flow is the suite's
// email one-time code, LOGIN ONLY — a Universal ID is created on the hub
// (app.unisim.co.uk), never here, so a typo cannot mint a stray account.
//
//   POST /auth/v1/otp     { email, create_user: false }      → code emailed
//   POST /auth/v1/verify  { email, token, type: 'email' }    → session
//   POST /auth/v1/token?grant_type=refresh_token             → fresh session
//   POST /auth/v1/logout                                     → ends it server-side
//
// ⚠️ **Signing in is entirely optional and changes nothing about how a
// connection works.** Screens is a LAN tool that must keep working with no
// route out: everything here fails quietly, and a signed-out client behaves
// exactly as it always has.
//
// ⚠️ **The PIN is never involved.** It is the Noise pre-shared key, it has
// never been written to disk by any client, and `screens_saved_hosts`
// (migration 0178) has no column for it. Sync carries the address, the
// machine's name and OS, and your nickname for it — nothing else.
//
// The session is kept in localStorage under this app's own key, deliberately
// NOT the suite's shared one: this page is served over plain http:// from a
// host on a LAN, which is a different origin from every other suite app and
// cannot share their cookie anyway.

const SUPABASE_URL = 'https://rygfxgalojojppxmhddo.supabase.co'
const SUPABASE_ANON =
  'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6InJ5Z2Z4Z2Fsb2pvanBweG1oZGRvIiwicm9sZSI6ImFub24iLCJpYXQiOjE3Nzg3NTY4MjUsImV4cCI6MjA5NDMzMjgyNX0.hLy_vt9vY_rdPKF3nL32yAuMCD604E3CH5VM7D7CaNE'

const SESSION_KEY = 'universal-screens.session'

/** Read the stored session, or null. */
function stored() {
  try {
    const raw = JSON.parse(localStorage.getItem(SESSION_KEY) ?? 'null')
    return raw && raw.refresh_token ? raw : null
  } catch {
    return null
  }
}

function store(session) {
  try {
    if (session) localStorage.setItem(SESSION_KEY, JSON.stringify(session))
    else localStorage.removeItem(SESSION_KEY)
  } catch {
    /* private mode — the session then lasts as long as the page does */
  }
  memory = session
  for (const fn of listeners) fn(session)
}

let memory = stored()
const listeners = new Set()

/** Call `fn` whenever the signed-in account changes. Returns an unsubscribe. */
export function onAccountChange(fn) {
  listeners.add(fn)
  return () => listeners.delete(fn)
}

async function auth(path, body, query = '') {
  const res = await fetch(`${SUPABASE_URL}/auth/v1/${path}${query}`, {
    method: 'POST',
    headers: { apikey: SUPABASE_ANON, 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  })
  const text = await res.text()
  const data = text ? JSON.parse(text) : {}
  if (!res.ok) throw new Error(data.error_description || data.msg || data.message || `HTTP ${res.status}`)
  return data
}

function shape(data) {
  if (!data?.access_token) return null
  return {
    access_token: data.access_token,
    refresh_token: data.refresh_token,
    // A minute of slack, so a request is never sent with a token about to expire.
    expires_at: Date.now() + Math.max(0, (data.expires_in ?? 3600) - 60) * 1000,
    email: data.user?.email ?? memory?.email ?? '',
    user_id: data.user?.id ?? memory?.user_id ?? '',
  }
}

/** Email a 6-digit code. Rejects with a readable message. */
export async function sendCode(email) {
  try {
    await auth('otp', { email: email.trim(), create_user: false })
  } catch (err) {
    // GoTrue answers an unknown address with "Signups not allowed for otp"
    // when create_user is false. Say what that means for this product.
    if (/signups? not allowed/i.test(String(err.message))) {
      throw new Error('No Universal ID for that email. Create one free at app.unisim.co.uk, then sign in here.')
    }
    throw err
  }
}

/** Exchange the emailed code for a session. */
export async function verifyCode(email, token) {
  const session = shape(await auth('verify', { email: email.trim(), token: token.trim(), type: 'email' }))
  if (!session) throw new Error('That code was not accepted.')
  store(session)
  return session
}

/** The signed-in account, refreshing the token when it is due. Null signed out. */
export async function session() {
  const current = memory
  if (!current) return null
  if (current.expires_at > Date.now()) return current
  try {
    const next = shape(await auth('token', { refresh_token: current.refresh_token }, '?grant_type=refresh_token'))
    if (!next) throw new Error('no session')
    store(next)
    return next
  } catch {
    // A refused refresh means the session is gone for good (signed out
    // elsewhere, account deleted, token revoked) — don't keep a dead one.
    store(null)
    return null
  }
}

/** Who is signed in, without touching the network. */
export function account() {
  return memory ? { email: memory.email, id: memory.user_id } : null
}

export async function signOut() {
  const current = memory
  store(null)
  if (!current) return
  try {
    await fetch(`${SUPABASE_URL}/auth/v1/logout`, {
      method: 'POST',
      headers: { apikey: SUPABASE_ANON, Authorization: `Bearer ${current.access_token}` },
    })
  } catch {
    /* the local session is already gone, which is what the user asked for */
  }
}

/** An authorised REST call, or null when signed out / the call fails. */
async function rest(path, { method = 'GET', body, prefer } = {}) {
  const s = await session()
  if (!s) return null
  try {
    const res = await fetch(`${SUPABASE_URL}/rest/v1/${path}`, {
      method,
      headers: {
        apikey: SUPABASE_ANON,
        Authorization: `Bearer ${s.access_token}`,
        'Content-Type': 'application/json',
        ...(prefer ? { Prefer: prefer } : {}),
      },
      ...(body ? { body: JSON.stringify(body) } : {}),
    })
    if (!res.ok) return null
    const text = await res.text()
    return text ? JSON.parse(text) : []
  } catch {
    return null
  }
}

/**
 * Merge this device's saved machines with the account's, newest-connected
 * winning per address, and write back whatever the other side was missing.
 *
 * Returns the merged list (most recent first), or null when signed out or
 * offline — the caller then just keeps showing what it had.
 *
 * ⚠️ There are no tombstones: deleting a saved machine while signed OUT only
 * affects that device, and the row comes back on the next sync. Deleting while
 * signed in calls `forget()` and is gone everywhere. The connect screen says
 * so rather than pretending otherwise.
 */
export async function syncSaved(local) {
  const remote = await rest('screens_saved_hosts?select=addr,hostname,os,custom_name,last_connected')
  if (remote === null) return null

  const merged = new Map()
  for (const r of remote) {
    merged.set(r.addr, {
      addr: r.addr,
      hostname: r.hostname ?? '',
      os: r.os ?? '',
      customName: r.custom_name ?? '',
      lastConnected: Date.parse(r.last_connected) || 0,
    })
  }
  const push = []
  for (const l of local) {
    const r = merged.get(l.addr)
    if (!r || l.lastConnected > r.lastConnected) {
      merged.set(l.addr, l)
      push.push(l)
    }
  }
  // Rows this device has never seen are already in `merged` and need no write.
  if (push.length) {
    const s = await session()
    await rest('screens_saved_hosts?on_conflict=user_id,addr', {
      method: 'POST',
      prefer: 'resolution=merge-duplicates',
      body: push.map((h) => ({
        user_id: s.user_id,
        addr: h.addr,
        hostname: h.hostname ?? '',
        os: h.os ?? '',
        custom_name: h.customName ?? '',
        last_connected: new Date(h.lastConnected || Date.now()).toISOString(),
      })),
    })
  }
  return [...merged.values()].sort((a, b) => b.lastConnected - a.lastConnected)
}

/** Save (or refresh) one machine against the account. No-op signed out. */
export async function remember(host) {
  const s = await session()
  if (!s) return
  await rest('screens_saved_hosts?on_conflict=user_id,addr', {
    method: 'POST',
    prefer: 'resolution=merge-duplicates',
    body: [{
      user_id: s.user_id,
      addr: host.addr,
      hostname: host.hostname ?? '',
      os: host.os ?? '',
      custom_name: host.customName ?? '',
      last_connected: new Date(host.lastConnected || Date.now()).toISOString(),
    }],
  })
}

/** Drop one machine from the account. No-op signed out. */
export async function forget(addr) {
  if (!memory) return
  await rest(`screens_saved_hosts?addr=eq.${encodeURIComponent(addr)}`, { method: 'DELETE' })
}

/** The access token for the presence beat, so a signed-in visit counts as the
 *  account rather than as this browser. Null signed out. */
export async function accessToken() {
  const s = await session()
  return s?.access_token ?? null
}
